grafanalib/opentsdb.py

```diff
@@ -0,0 +1,156 @@
+import attr
+from attr.validators import instance_of
+
+# OpenTSDB aggregators
+OTSDB_AGG_AVG = "avg"
+OTSDB_AGG_COUNT = "count"
+OTSDB_AGG_DEV = "dev"
+OTSDB_AGG_EP50R3 = "ep50r3"
+OTSDB_AGG_EP50R7 = "ep50r7"
+OTSDB_AGG_EP75R3 = "ep75r3"
+OTSDB_AGG_EP75R7 = "ep75r7"
+OTSDB_AGG_EP90R3 = "ep90r3"
+OTSDB_AGG_EP90R7 = "ep90r7"
+OTSDB_AGG_EP95R3 = "ep95r3"
+OTSDB_AGG_EP95R7 = "ep95r7"
+OTSDB_AGG_EP99R3 = "ep99r3"
+OTSDB_AGG_EP99R7 = "ep99r7"
+OTSDB_AGG_EP999R3 = "ep999r3"
+OTSDB_AGG_EP999R7 = "ep999r7"
+OTSDB_AGG_FIRST = "first"
+OTSDB_AGG_LAST = "last"
+OTSDB_AGG_MIMMIN = "mimmin"
+OTSDB_AGG_MIMMAX = "mimmax"
+OTSDB_AGG_MIN = "min"
+OTSDB_AGG_MAX = "max"
+OTSDB_AGG_NONE = "none"
+OTSDB_AGG_P50 = "p50"
+OTSDB_AGG_P75 = "p75"
+OTSDB_AGG_P90 = "p90"
+OTSDB_AGG_P95 = "p95"
+OTSDB_AGG_P99 = "p99"
+OTSDB_AGG_P999 = "p999"
+OTSDB_AGG_SUM = "sum"
+OTSDB_AGG_ZIMSUM = "zimsum"
+
+OTSDB_DOWNSAMPLING_FILL_POLICIES = ('none', 'nan', 'null', 'zero')
+OTSDB_DOWNSAMPLING_FILL_POLICY_DEFAULT = 'none'
+
+OTSDB_QUERY_FILTERS = (
+    'literal_or', 'ilteral_or', 'not_literal_or',
+    'not_iliteral_or', 'wildcard', 'iwildcard', 'regexp')
+OTSDB_QUERY_FILTER_DEFAULT = 'literal_or'
+
+
+def valid_opentsdb_downsample_fill_policy(instance, attribute, value):
+    if value not in OTSDB_DOWNSAMPLING_FILL_POLICIES:
+        raise ValueError("{attr} shuld be one of {choice}".format(
+            attr=attribute, choice=OTSDB_DOWNSAMPLING_FILL_POLICIES))
+
+
+def valid_opentsdb_filter(instance, attribute, value):
+    if value not in OTSDB_QUERY_FILTERS:
+        raise ValueError("{attr} shuld be one of {choice}".format(
+            attr=attribute, choice=OTSDB_QUERY_FILTERS))
+
+
+@attr.s
+class OpenTSDBFilter(object):
+
+    value = attr.ib()
+    tag = attr.ib()
+    type = attr.ib(
+        default=OTSDB_QUERY_FILTER_DEFAULT,
+        validator=valid_opentsdb_filter)
+    groupBy = attr.ib(default=False, validator=instance_of(bool))
+
+    def to_json_data(self):
+        return {
+            'filter': self.value,
+            'tagk': self.tag,
```

Comment:
No, I don't know for what reason, but they name this field `tagk` in api.
Probably, because previously opentsb had another filtering mechanism based on tags and they wanted to distinguish new type of filtering from old one.